jax/experimental/pjit.py

```diff
@@ -0,0 +1,211 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import OrderedDict
+from typing import (Callable, Iterable, Optional, Tuple, Union)
+from warnings import warn
+
+from . import maps
+from .. import core
+from .. import linear_util as lu
+from ..api import _check_callable, _check_arg
+from ..api_util import (argnums_partial, flatten_axes, flatten_fun_nokwargs,
+                        _ensure_index_tuple)
+from ..interpreters import ad
+from ..interpreters import pxla
+from ..interpreters import xla
+from ..lib import xla_bridge as xb
+from ..lib import xla_client as xc
+from ..tree_util import tree_flatten, tree_unflatten
+from .._src.util import (extend_name_stack, HashableFunction, safe_zip,
+                         wrap_name, wraps)
+xops = xc._xla.ops
+
+# TODO(skye): better name than mesh_axes
+def pjit(fun: Callable,
+         in_mesh_axes,
```

Comment:
Those are supposed to be like `xmap`'s `in_axes`, but the leaves are `PartitionSpec`s of mesh axis names and `None`s, right?

Reply:
I would maybe call these arguments `in_axis_resources` and `out_axis_resources` for max consistency with `xmap`?

Reply:
Yeah ok. I actually find the term `axis_resources` kinda confusing, but I'll use it here for consistency and maybe we can reconsider it for both `xmap` and `pjit` at some point.

[bikeshedding on naming below]
After studying the xmap docstring, I think the "axis" in `axis_resources` refers to the named array axes? Elsewhere in the docstring, the mesh axes are called "resource axes". So I think `axis_resources` is short for `axis_resource_axes`? If so, IMO renaming to `resource_axes` is slightly clearer.